### PR TITLE
Implement Exp Charm and unevolved Exp multipliers + Exp formula fixes

### PR DESCRIPTION
**Pull request by kittenchilly**
## Description
Gives the Exp charm its functionality, boosting all exp by 1.5x if you have it in your inventory.
Also adds the missing ~1.2x exp multiplier with a config for if the mon is at or past the level where it would be able to evolve, but it has not.

Alongside these it fixes:
Boosted Exp from traded mon accidentally being applied twice
Affection multiplier being exactly x1.2 instead of 4915/4096 like it is on cart
Trainer multiplier being applied with other multipliers when it should be before the exp share scaling

## **Discord contact info**
kittenchilly

### CODE REVIEW

File: src/pokemon.c [hunk 1]
```diff
@@ -6770,6 +6770,131 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, s
...
+    u32 personality = GetMonData(mon, MON_DATA_PERSONALITY, 0);
+    u8 level = GetMonData(mon, MON_DATA_LEVEL, 0);
+    u16 upperPersonality = personality >> 16;
+    
+    for (i = 0; i < EVOS_PER_MON; i++)
+    {
+        switch (gEvolutionTable[species][i].method)
+        {
+        case EVO_LEVEL_DAY:
+            RtcCalcLocalTime();
```
Comment (AsparagusEduardo) [inline src/pokemon.c:6786]: This implies that the multiplier only works for these when the time of day is correct, which Bulbapedia's description doesn't imply.

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -4093,6 +4093,9 @@ static void Cmd_getexp(void)
                 calculatedExp = gSpeciesInfo[gBattleMons[gBattlerFainted].species].expYield * gBattleMons[gBattlerFainted].level / 7;
             #endif
 
+            if (B_TRAINER_EXP_MULTIPLIER <= GEN_7 && gBattleTypeFlags & BATTLE_TYPE_TRAINER)
+                calculatedExp = (calculatedExp * 150) / 100;
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:4097]: Why did you move this out of `ApplyExperienceMultipliers`?

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_script_commands.c (hunk 2 above)
Comment (kittenchilly) [reply src/battle_script_commands.c:4097]: If you look at the formula that includes this (a here), it is done before the exp share multiplier. Moving it here is more accurate.
image

File: src/pokemon.c (hunk 1 above)
Comment (kittenchilly) [reply src/pokemon.c:6786]: I was under the impression this multiplier is for helping you evolve pokemon that only evolve when leveling up in a specific way, if you missed it before. So it only giving the boost during that specific time makes sense. Not really sure though, so I guess removing the requirement is safest

Comment (AsparagusEduardo) [reply src/pokemon.c:6786]: Please confirm how it works before doing so.

Comment (kittenchilly) [reply src/pokemon.c]: Confirmed it only checks level and no other condition, apparently.

**Comment by kittenchilly** [comment]
I believe this PR is finished, and should be reviewable again

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
I thought that this review went through but it was still pending.

**Review by AsparagusEduardo — APPROVED** [review]
